### PR TITLE
add signing spec

### PR DESCRIPTION
**Pull request by ebuchman**
Ref https://github.com/tendermint/tendermint/issues/2820

Spec for those implementing validator signers.

* [ ] Updated all relevant documentation in docs
* [ ] Updated all code comments where relevant
* [ ] Wrote tests
* [ ] Updated CHANGELOG_PENDING.md

### CODE REVIEW

**Comment by ebuchman** [comment]
I'm noticing now that we somehow managed to mangle the order of BlockID and Timestamp in the Canonical Proposal and Vote :(

I hope we can still change that, but maybe it's too late

File: docs/spec/consensus/signing.md [hunk 1]
```diff
@@ -0,0 +1,205 @@
...
+and to only sign votes and proposals corresponding to that ChainID.
+
+## BlockID
+
+BlockID is the structure used to represent the block:
+
+```
+type BlockID struct {
+	Hash	[]byte
+	Header	PartSetHeader
```
Comment (Liamsi) [inline docs/spec/consensus/signing.md]: The var in the code is still named `PartsHeader`. Could we name the vars here (in the spec) and in the implementation consistently?

File: docs/spec/consensus/signing.md [hunk 2]
```diff
@@ -0,0 +1,205 @@
...
+}
+
+type PartSetHeader struct {
+	Hash  []byte
+	Total int
+}
+```
+
+To be included in a valid vote or proposal, BlockID must either represent a `nil` block, or a complete one.
+We introduce two methods, `BlockID.IsNil()` and `BlockID.IsComplete()` for these cases, respectively.
```
Comment (Liamsi) [inline docs/spec/consensus/signing.md]: Introducing these methods in the go-code as well would make the validation more readable.

**Comment by Liamsi** [comment]
Oh no. The question is, if we'd need to be backwards compatible in some way? If not, it should be fairly easy to change that.

**Comment by ebuchman** [comment]
> The question is, if we'd need to be backwards compatible in some way? If not, it should be fairly easy to change that. 

We need to make a non-backwards compatible release soon (hopefully the last one for some time), so let's include it in there. I opened an issue: https://github.com/tendermint/tendermint/issues/3077

**Pull request merged**